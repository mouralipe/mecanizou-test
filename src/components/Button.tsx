interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  children: React.ReactNode;
  variant?: 'default' | 'icon' | 'outline';
}

export function Button({
  children,
  variant = 'default',
  ...props
}: ButtonProps) {
  if (variant === 'icon') {
    return (
      <button className={`cursor-pointer`} {...props}>
        {children}
      </button>
    );
  }

  if (variant === 'outline') {
    return (
      <button
        className="cursor-pointer px-3 py-2 border border-gray-300 dark:border-gray-600 rounded-md disabled:opacity-50"
        {...props}
      >
        {children}
      </button>
    );
  }

  return (
    <button
      className={`bg-blue-500 px-3 py-2 border border-gray-300 dark:border-gray-600 text-white rounded-md cursor-pointer`}
      {...props}
    >
      {children}
    </button>
  );
}
