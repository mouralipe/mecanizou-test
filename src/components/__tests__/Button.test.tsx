import { fireEvent, render, screen } from '@testing-library/react';
import { Button } from '../Button';

describe('Button Component', () => {
  it('renders with children', () => {
    render(<Button>Click me</Button>);
    expect(
      screen.getByRole('button', { name: /click me/i })
    ).toBeInTheDocument();
  });

  it('renders default variant correctly', () => {
    render(<Button>Default</Button>);
    const button = screen.getByRole('button');
    expect(button).toHaveClass(
      'bg-blue-500',
      'px-3',
      'py-2',
      'text-white',
      'rounded-md'
    );
  });

  it('renders outline variant correctly', () => {
    render(<Button variant="outline">Outline</Button>);
    const button = screen.getByRole('button');
    expect(button).toHaveClass(
      'px-3',
      'py-2',
      'border',
      'border-gray-300',
      'rounded-md'
    );
    expect(button).not.toHaveClass('bg-blue-500', 'text-white');
  });

  it('renders icon variant correctly', () => {
    render(<Button variant="icon">🎯</Button>);
    const button = screen.getByRole('button');
    expect(button).toHaveClass('cursor-pointer');
    expect(button).not.toHaveClass('bg-blue-500', 'px-3', 'py-2');
  });

  it('handles click events', () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Click me</Button>);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('handles disabled state', () => {
    render(<Button disabled>Disabled</Button>);
    const button = screen.getByRole('button');
    expect(button).toBeDisabled();
  });

  it('applies custom props correctly', () => {
    render(
      <Button data-testid="custom-button" type="submit">
        Submit
      </Button>
    );
    const button = screen.getByTestId('custom-button');
    expect(button).toHaveAttribute('type', 'submit');
  });

  it('matches snapshot for default variant', () => {
    const { container } = render(<Button>Default Button</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for outline variant', () => {
    const { container } = render(
      <Button variant="outline">Outline Button</Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for icon variant', () => {
    const { container } = render(<Button variant="icon">🎯</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
