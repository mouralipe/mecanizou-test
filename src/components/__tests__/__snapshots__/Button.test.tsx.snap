// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button Component matches snapshot for default variant 1`] = `
<button
  class="bg-blue-500 px-3 py-2 border border-gray-300 dark:border-gray-600 text-white rounded-md cursor-pointer"
>
  Default Button
</button>
`;

exports[`Button Component matches snapshot for icon variant 1`] = `
<button
  class="cursor-pointer"
>
  ðŸŽ¯
</button>
`;

exports[`Button Component matches snapshot for outline variant 1`] = `
<button
  class="cursor-pointer px-3 py-2 border border-gray-300 dark:border-gray-600 rounded-md disabled:opacity-50"
>
  Outline Button
</button>
`;
